### PR TITLE
Feat/refactor dispatcher

### PR DESCRIPTION
**Pull request by mariogiampieri**
## Description

- Begin porting over events bound to map event types, primarily `Hover` and extensions in the original codebase. Map events are now bound to handler functions in `mapEvents.ts`. 
- Collect and re-organize top-level `Districtr` object properties / components library `dispatch` as elements of zustand state. In so doing, rename / replace `zoneStore` as `mapStore` to be more general. Notes on the status of this process are in docs/refactor_notes/reducer_dispatch_notes.md (maybe this should live elsewhere?)
- Begin implementing `react-query` to manage server transactions. Specifically, begin using `useMutation` to generate hook with which to `post` to database. Per the docs, mutations are best-suited for cases when transactions update server-side code, as opposed to queries, which primarily receive data from the server. A generic implementation lives in `apiHandlers.ts` and is invoked as e.g. `saveMap.mutate(map-object-var-name)`. This has not yet been tested on db endpoints. 
- **Most notably**, the first two points above combined provide the click-drag-paint functionality that the current app does, along with a reliable hover state. 
- Still some optimization to do in terms of updating map, store, and server. 

## Reviewers

- Primary: @raphaellaude 
- Secondary: @bailliekova 

## Checklist

- [ ] Added/Updated related documentation (if applicable).
- [ ] Added/Updated related unit tests (if applicable).

### CODE REVIEW

**Comment by mariogiampieri** [comment]
Next step will be to actually save / update assignments, to come

File: app/src/app/store/mapStore.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+   * @type {Set<string>}
+   */
+  accumulatedGeoids: new Set<string>(),
+  /**
+   *
+   * @param zone - identifier for the zone assignment
+   * @param geoids - Set of geoids to assign to the zone; members of the set will be assigned to the zone
+   * @returns updated dict of geoid: zone assignments
+   */
+  setZoneAssignments: (zone: Zone, geoids: Set<GEOID>) =>
```
Comment (raphaellaude) [inline app/src/app/store/mapStore.ts:78]: Sadly for my pet theory it seems like maplibregl [doesn't support easily querying feature state](https://docs.mapbox.com/style-spec/reference/layers/#filter) for more than 1 id? So we'll have to have this temp store

File: app/src/app/utils/events/mapEvents.ts [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    if (activeTool === "brush") {
+      mapStore.setIsPainting(true);
+      return;
+    } else if (activeTool === "eraser") {
+      // erase features tbd
+    }
+  }
+};
+
+export const handleMapMouseEnter = (
```
Comment (raphaellaude) [inline app/src/app/utils/events/mapEvents.ts:85]: These first four are just unimplemented for the time being right?

File: app/src/app/utils/events/mapEvents.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  if (!isPainting) {
+    HighlightFeature(selectedFeatures, map, hoverFeatureIds);
+  } else if (activeTool === "brush" && isPainting) {
+    /**
+     * @todo
+     * what we really want is to set map feature state here,
+     * and then update the store with the new assignments when
+     * we mouseup, to avoid unnecessary rerenders and state updates.
+     * this should reduce the bottleneck from debouncing
+     * */
```
Comment (raphaellaude) [inline app/src/app/utils/events/mapEvents.ts:131]: Ooo nice I like that

**Review by raphaellaude — CHANGES_REQUESTED** [review]
Super impressive bit of work.

I fetched the PR and played around with it. Definitely feels pretty snippy at smaller scales but a bit laggy for the whole map. This will be better once we implement splitting of course since most maps won't be full block maps.

Only one blocking comment: would love to refactor `uuid` as `document_id` sooner than later.

File: app/src/app/api/apiHandlers.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
+import { useMutation } from "@tanstack/react-query";
+import axios from "axios";
+import type { Map } from "maplibre-gl";
+import { useMapStore } from "@/app/store/mapStore";
+import { use } from "react";
```
[the remark this answers is not in the document]
Comment (mariogiampieri) [reply app/src/app/api/apiHandlers.ts:5]: yep def, legacy import

File: app/src/app/api/apiHandlers.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      console.log(`Mutation ${context.id} successful!`, data);
+    },
+    onSettled: (data, error, variables, context) => {
+      // fires regardless of error or success
+      console.log(`Optimistic update with id ${context?.id} settled: `);
+      if (error) {
+        console.log("Error: ", error);
+      }
+      if (data) {
+        useMapStore.setState({ uuid: data.data });
```
[the remark this answers is not in the document]
Comment (mariogiampieri) [reply app/src/app/api/apiHandlers.ts:70]: updated here and in store

File: app/src/app/api/apiHandlers.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+/**
+ * Save map data to the server.
+ * @param mapObject - Map, the map object to save. In this case, the entire maplibre map object.
+ * @returns Promise
+ */
+const postMapObject: (mapObject: Map) => Promise<responseObject> = async (
+  mapObject: Map
+) => {
+  // return axios.post("/saveMap", mapObject);
```
[the remark this answers is not in the document]
Comment (mariogiampieri) [reply app/src/app/api/apiHandlers.ts:90]: that would be great- this is overly generic right now so looking forward to refining what is actually being sent over the wire

File: app/src/app/store/mapStore.ts (hunk 1 above)
Comment (mariogiampieri) [reply app/src/app/store/mapStore.ts:78]: bummer, but copy that

File: app/src/app/utils/events/mapEvents.ts (hunk 2 above)
Comment (mariogiampieri) [reply app/src/app/utils/events/mapEvents.ts:85]: exactly, right now these are just placeholders

File: app/src/app/api/apiHandlers.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mariogiampieri) [reply app/src/app/api/apiHandlers.ts:5]: `use` was removed; `Map` will also be removed when we refine what's being passed over the wire (right now this is a generic placeholder)

File: app/src/app/utils/events/mapEvents.ts (hunk 3 above)
Comment (mariogiampieri) [reply app/src/app/utils/events/mapEvents.ts:131]: referred to in #7, will treat that as a discrete bit of work after merging this in

File: app/src/app/constants/types.ts [hunk 7]
```diff
@@ -1,7 +1,31 @@
...
 export type Zone = number;
 
 export type GEOID = string;
 
-// create a dict of zone: [geoid]
 export type ZoneDict = Map<GEOID, Zone>;
+
+export type ActiveTool = "pan" | "brush" | "eraser"; // others?
+
+export type SpatialUnit = "county" | "tract" | "block"; // others?
```
[the remark this answers is not in the document]
Comment (mariogiampieri) [reply app/src/app/constants/types.ts:11]: added block group- i feel slow but what's vdt?

[the remark this answers is not in the document]
Comment (raphaellaude) [reply app/src/app/constants/types.ts]: Voting districts!

**Review by raphaellaude — APPROVED** [review]
Lgtm

File: app/src/app/constants/types.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (mariogiampieri) [reply app/src/app/constants/types.ts]: of course!

**Review by raphaellaude — APPROVED** [review]

**Pull request merged**